lib/src/pub.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright 2019 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:grinder/grinder.dart';
+import 'package:path/path.dart' as p;
+
+import 'utils.dart';
+
+/// The entire contents of pub's `~/.pub-cache/credentials.json` file.
+///
+/// **Do not check this in directly.** This should only come from secure
+/// sources.
+///
+/// By default this comes from the `PUB_CREDENTIALS` environment variable.
+String get pubCredentials {
+  _pubCredentials ??= Platform.environment["PUB_CREDENTIALS"];
+  if (_pubCredentials != null) return _pubCredentials;
+
+  fail("pkg.pubCredentials must be set to deploy to pub.");
+}
+
+set pubCredentials(String value) => _pubCredentials = value;
+String _pubCredentials;
+
+/// The path in which pub expects to find its credentials file.
+final String _credentialsPath = () {
+  String cacheDir;
+  if (Platform.environment.containsKey('PUB_CACHE')) {
+    cacheDir = Platform.environment['PUB_CACHE'];
+  } else if (Platform.isWindows) {
+    var appData = Platform.environment['APPDATA'];
+    cacheDir = p.join(appData, 'Pub', 'Cache');
+  } else {
+    cacheDir = p.join(Platform.environment['HOME'], '.pub-cache');
```

Comment:
nit: Add a comment here linking back to the source of truth where the paths are defined for windows and other platforms.

Reply:
Done.